app/models/concerns/category_hashtag.rb

```diff
@@ -51,15 +50,15 @@ def query_from_cached_categories(category_slugs, cached_categories)
           # by its slug then find the child by its slug and its parent's
           # ID to make sure they match.
           if child_slug.present?
-            parent_category = cached_categories.find { |cat| cat[:slug].downcase == parent_slug }
+            parent_category = categories.find { |cat| cat.slug.downcase == parent_slug }
             if parent_category.present?
-              cached_categories.find do |cat|
-                cat[:slug].downcase == child_slug && cat[:parent_category_id] == parent_category[:id]
+              categories.find do |cat|
+                cat.slug.downcase == child_slug && cat.parent_category_id == parent_category.id
               end
             end
           else
-            cached_categories.find do |cat|
-              cat[:slug].downcase == parent_slug && cat[:parent_category_id].nil?
+            categories.find do |cat|
+              cat.slug.downcase == parent_slug && cat.parent_category_id.nil?
```

Comment:
Perhaps we could lift the second condition into `Category#top_level?` or similar? I see checks for `parent_category_id` in other parts of the code, especially to check the inverse, which could be `Category#subcategory?` or similar, defined in terms of `Category#top_level?`.

Reply:
Oh yep I see what you mean, that would be useful, I will just add `top_level?` in this particular commit